receiver/oktareceiver/logs.go

```diff
@@ -0,0 +1,235 @@
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+type oktaLogsReceiver struct {
+	cfg      Config
+	client   httpClient
+	consumer consumer.Logs
+	logger   *zap.Logger
+	nextURL  string
+	cancel   context.CancelFunc
+	wg       *sync.WaitGroup
+}
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+	CloseIdleConnections()
+}
+
+// newOktaLogsReceiver returns a newly configured oktaLogsReceiver
+func newOktaLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) (*oktaLogsReceiver, error) {
+	return &oktaLogsReceiver{
+		cfg:      *cfg,
+		client:   http.DefaultClient,
+		consumer: consumer,
+		logger:   logger,
+		wg:       &sync.WaitGroup{},
+	}, nil
+}
+
+func (r *oktaLogsReceiver) Start(_ context.Context, _ component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.wg.Add(1)
+	go r.startPolling(ctx)
+	return nil
+}
+
+func (r *oktaLogsReceiver) startPolling(ctx context.Context) {
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *oktaLogsReceiver) poll(ctx context.Context) error {
+	logEvents := r.getLogs(ctx)
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := r.processLogEvents(observedTime, logEvents)
+	if logs.LogRecordCount() > 0 {
+		if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *oktaLogsReceiver) getLogs(ctx context.Context) []*okta.LogEvent {
+	var logs []*okta.LogEvent
+	var reqURL string
+	pollTime := time.Now().UTC()
+
+	// get logs until there isn't any overflow OR we get logs published after initial pollTime
+	for {
+		if r.nextURL == "" {
+			reqURL = "https://" + r.cfg.Domain + "/api/v1/logs"
+		} else {
+			reqURL = r.nextURL
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
+		if err != nil {
+			r.logger.Error("error creating okta api request", zap.Error(err))
+			break
+		}
+
+		// add query params to the first polling request
+		if r.nextURL == "" {
+			query := req.URL.Query()
+			query.Add("since", pollTime.Add(-r.cfg.PollInterval).Format(OktaTimeFormat))
+			query.Add("limit", strconv.Itoa(oktaMaxLimit))
+			req.URL.RawQuery = query.Encode()
+		}
+
+		req.Header.Add("Authorization", "SSWS "+string(r.cfg.APIToken))
+
+		res, err := r.client.Do(req)
+		if err != nil {
+			r.logger.Error("error performing okta api request", zap.Error(err))
+			break
+		}
+
+		if res.Body != nil {
+			defer res.Body.Close()
+		}
+
+		if res.StatusCode != http.StatusOK {
+			r.logger.Error("okta logs endpoint returned non-200 statuscode: " + res.Status)
+			break
+		}
+
+		r.setNextLink(res)
```

Comment:
Kinda nit, but I would move this `r.setNextLink(res)` call to the end of this for loop. The result of this func doesn't seem to have an impact on anything happening after it where it's placed now.